shared-module/audiocore/__init__.h

```diff
@@ -31,22 +31,45 @@
 #include <stdint.h>
 
 #include "py/obj.h"
+#include "py/proto.h"
 
 typedef enum {
     GET_BUFFER_DONE,            // No more data to read
     GET_BUFFER_MORE_DATA,       // More data to read.
     GET_BUFFER_ERROR,           // Error while reading data.
 } audioio_get_buffer_result_t;
 
-uint32_t audiosample_sample_rate(mp_obj_t sample_obj);
-uint8_t audiosample_bits_per_sample(mp_obj_t sample_obj);
-uint8_t audiosample_channel_count(mp_obj_t sample_obj);
-void audiosample_reset_buffer(mp_obj_t sample_obj, bool single_channel, uint8_t audio_channel);
-audioio_get_buffer_result_t audiosample_get_buffer(mp_obj_t sample_obj,
+typedef uint32_t (*audiosample_sample_rate_fun)(void* sample_obj);
+typedef uint8_t (*audiosample_bits_per_sample_fun)(void* sample_obj);
+typedef uint8_t (*audiosample_channel_count_fun)(void* sample_obj);
+typedef void (*audiosample_reset_buffer_fun)(void* sample_obj);
+typedef audioio_get_buffer_result_t (*audiosample_get_buffer_fun)(void* sample_obj,
+        bool single_channel, uint8_t channel, uint8_t** buffer,
+        uint32_t* buffer_length);
+typedef void (*audiosample_get_buffer_structure_fun)(void* sample_obj,
+        bool single_channel, bool* single_buffer,
+        bool* samples_signed, uint32_t *max_buffer_length,
+        uint8_t* spacing);
+
+typedef struct _audiosample_p_t {
+    MP_PROTOCOL_HEAD // MP_QSTR_protocol_audiosample
+    audiosample_sample_rate_fun sample_rate;
+    audiosample_bits_per_sample_fun bits_per_sample;
+    audiosample_channel_count_fun channel_count;
+    audiosample_reset_buffer_fun reset_buffer;
+    audiosample_get_buffer_fun get_buffer;
+    audiosample_get_buffer_structure_fun get_buffer_structure;
+} audiosample_p_t;
+
+uint32_t audiosample_sample_rate(void* sample_obj);
+uint8_t audiosample_bits_per_sample(void* sample_obj);
+uint8_t audiosample_channel_count(void* sample_obj);
+void audiosample_reset_buffer(void* sample_obj, bool single_channel, uint8_t audio_channel);
+audioio_get_buffer_result_t audiosample_get_buffer(void* sample_obj,
                                                    bool single_channel,
                                                    uint8_t channel,
                                                    uint8_t** buffer, uint32_t* buffer_length);
-void audiosample_get_buffer_structure(mp_obj_t sample_obj, bool single_channel,
+void audiosample_get_buffer_structure(void* sample_obj, bool single_channel,
```

Comment:
Why switch to void* here?

Reply:
this bit should probably be reverted, I tried several approaches to putting the pointers into the protocol object (which has to refer to a basic type) and I think this reflects an intermediate approach of void*-typing at the interface.  I'll remove this.  Did you catch more than one site where I did it, or is this the only one you saw?

Reply:
This is only where I noticed it.

Reply:
there were more, in the same file; but obscured by the way diff presented itself.  I fixed them all.